command-line/lib/Options/Applicative/Script.hs

```diff
@@ -9,30 +9,77 @@ module Options.Applicative.Script
     (
     -- ** Applicative Parser
       scriptArg
+    , stakeCredentialArg
+    , StakeCredential (..)
     ) where
 
 import Prelude
 
+import Cardano.Address.Derivation
+    ( XPub )
 import Cardano.Script
-    ( Script (..) )
+    ( Script (..)
+    , ScriptError (..)
+    , ScriptHash
+    , scriptErrorToMsg
+    , scriptHashFromBytes
+    )
 import Cardano.ScriptParser
-    ( scriptParser )
+    ( scriptFromString )
+import Codec.Binary.Encoding
+    ( AbstractEncoding (..)
+    , detectEncoding
+    , fromBase16
+    , fromBase58
+    , fromBech32
+    )
+import Control.Applicative
+    ( (<|>) )
 import Options.Applicative
     ( Parser, argument, eitherReader, help, metavar )
-import Text.ParserCombinators.ReadP
-    ( readP_to_S )
+import Options.Applicative.Derivation
+    ( xpubArg )
+import System.IO.Extra
+    ( markCharsRedAtIndices )
 
-
-scriptFromString :: String -> Either String Script
-scriptFromString str =
-    case readP_to_S scriptParser str of
-         [(multisig,_rest)] ->
-                 Right multisig
-         _ ->
-             Left "Parsing of the script failed."
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
 
 scriptArg :: Parser Script
-scriptArg = argument (eitherReader scriptFromString) $ mempty
+scriptArg = argument (eitherReader scriptFromString') $ mempty
     <> metavar "SCRIPT"
     <> help
         "Script string."
+
+scriptFromString' :: String -> Either String Script
+scriptFromString' = maybe (Left errMsg) Right . scriptFromString
+  where
+    errMsg = scriptErrorToMsg MalformedScript
+
+scriptHashReader :: String -> Either String ScriptHash
+scriptHashReader str = do
+    bytes <- case detectEncoding str of
+        Just EBase16   -> fromBase16 (toBytes str)
+        Just EBech32{} -> fromBech32 markCharsRedAtIndices (toBytes str)
+        Just EBase58   -> fromBase58 (toBytes str)
+        Nothing        -> Left
+            "Couldn't detect input encoding? The key must be encoded as \
+            \base16, bech32 or base58."
+    case scriptHashFromBytes bytes of
+        Just scriptHash -> pure scriptHash
+        Nothing   -> Left
+            "Failed to convert bytes into a valid script hash."
+  where
+    toBytes = T.encodeUtf8 . T.pack
+
+scriptHashArg :: String -> Parser ScriptHash
+scriptHashArg helpDoc =
+    argument (eitherReader scriptHashReader) $ mempty
+        <> metavar "SCRIPT_HASH"
+        <> help helpDoc
+
+data StakeCredential = FromKey XPub | FromScriptHash ScriptHash
+    deriving (Show, Eq)
+
+stakeCredentialArg  :: String -> Parser StakeCredential
+stakeCredentialArg str = (FromScriptHash <$> scriptHashArg str) <|> (FromKey <$> xpubArg str)
```

Comment:
need to do something that when xpub is in str then `Failed to convert bytes into a valid script hash` is not thrown

Reply:
fixed in https://github.com/input-output-hk/cardano-addresses/pull/79/commits/1aee6f8ba433eef62452070b75a2fceee978752c